Build failure: python3{12,13}Packages.jwt

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

1. Build `python3Packages.jwt` or a package that depends on it (for example `devtoolbox`)

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/300645066

### Relevant log output

```console
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing pytest-check-hook
Using pytestCheckPhase
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/f439f7mla5dpdk6fkcrjz41dn8bd0j5h-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file "source/tox.ini"
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
substituteStream() in derivation python3.12-jwt-1.3.1: WARNING: '--replace' is deprecated, use --replace-{fail,warn,quiet}. (file 'setup.cfg')
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing setuptoolsBuildPhase
setup.py build flags: ''
/nix/store/v6zr0v80z1i73fv844am95blkwa01y7d-python3.12-setuptools-80.7.1/lib/python3.12/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
!!

        ********************************************************************************
        Please consider removing the following classifiers in favor of a SPDX license expression:

        License :: OSI Approved :: Apache Software License

        See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
        ********************************************************************************

!!
  self._finalize_license_expression()
running bdist_wheel
running build
running build_py
creating build/lib/jwt
copying jwt/jwa.py -> build/lib/jwt
copying jwt/utils.py -> build/lib/jwt
copying jwt/jwkset.py -> build/lib/jwt
copying jwt/__init__.py -> build/lib/jwt
copying jwt/exceptions.py -> build/lib/jwt
copying jwt/jwt.py -> build/lib/jwt
copying jwt/jwk.py -> build/lib/jwt
copying jwt/jws.py -> build/lib/jwt
copying jwt/py.typed -> build/lib/jwt
/nix/store/v6zr0v80z1i73fv844am95blkwa01y7d-python3.12-setuptools-80.7.1/lib/python3.12/site-packages/setuptools/_distutils/cmd.py:90: SetuptoolsDeprecationWarning: setup.py install is deprecated.
!!

        ********************************************************************************
        Please avoid running ``setup.py`` directly.
        Instead, use pypa/build, pypa/installer or other
        standards-based tools.

        By 2025-Oct-31, you need to update your project and remove deprecated calls
        or your builds will no longer be supported.

        See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
        ********************************************************************************

!!
  self.initialize_options()
installing to build/bdist.linux-x86_64/wheel
running install
running install_lib
creating build/bdist.linux-x86_64/wheel
creating build/bdist.linux-x86_64/wheel/jwt
copying build/lib/jwt/jwa.py -> build/bdist.linux-x86_64/wheel/./jwt
copying build/lib/jwt/utils.py -> build/bdist.linux-x86_64/wheel/./jwt
copying build/lib/jwt/jwkset.py -> build/bdist.linux-x86_64/wheel/./jwt
copying build/lib/jwt/__init__.py -> build/bdist.linux-x86_64/wheel/./jwt
copying build/lib/jwt/exceptions.py -> build/bdist.linux-x86_64/wheel/./jwt
copying build/lib/jwt/jwt.py -> build/bdist.linux-x86_64/wheel/./jwt
copying build/lib/jwt/py.typed -> build/bdist.linux-x86_64/wheel/./jwt
copying build/lib/jwt/jwk.py -> build/bdist.linux-x86_64/wheel/./jwt
copying build/lib/jwt/jws.py -> build/bdist.linux-x86_64/wheel/./jwt
running install_egg_info
running egg_info
creating jwt.egg-info
writing jwt.egg-info/PKG-INFO
writing dependency_links to jwt.egg-info/dependency_links.txt
writing requirements to jwt.egg-info/requires.txt
writing top-level names to jwt.egg-info/top_level.txt
writing manifest file 'jwt.egg-info/SOURCES.txt'
reading manifest file 'jwt.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'jwt.egg-info/SOURCES.txt'
Copying jwt.egg-info to build/bdist.linux-x86_64/wheel/./jwt-1.3.1-py3.12.egg-info
running install_scripts
creating build/bdist.linux-x86_64/wheel/jwt-1.3.1.dist-info/WHEEL
creating 'dist/jwt-1.3.1-py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
adding 'jwt/__init__.py'
adding 'jwt/exceptions.py'
adding 'jwt/jwa.py'
adding 'jwt/jwk.py'
adding 'jwt/jwkset.py'
adding 'jwt/jws.py'
adding 'jwt/jwt.py'
adding 'jwt/py.typed'
adding 'jwt/utils.py'
adding 'jwt-1.3.1.dist-info/licenses/LICENSE'
adding 'jwt-1.3.1.dist-info/METADATA'
adding 'jwt-1.3.1.dist-info/WHEEL'
adding 'jwt-1.3.1.dist-info/top_level.txt'
adding 'jwt-1.3.1.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
Finished executing setuptoolsBuildPhase
Running phase: installPhase
@nix { "action": "setPhase", "phase": "installPhase" }
Executing pypaInstallPhase
Successfully installed jwt-1.3.1-py3-none-any.whl
Finished executing pypaInstallPhase
Running phase: pythonOutputDistPhase
@nix { "action": "setPhase", "phase": "pythonOutputDistPhase" }
Executing pythonOutputDistPhase
Finished executing pythonOutputDistPhase
Running phase: fixupPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
shrinking RPATHs of ELF executables and libraries in /nix/store/1ykfhhlwy5ah9vr2dy4ns8a5fna7wbnq-python3.12-jwt-1.3.1
checking for references to /build/ in /nix/store/1ykfhhlwy5ah9vr2dy4ns8a5fna7wbnq-python3.12-jwt-1.3.1...
patching script interpreter paths in /nix/store/1ykfhhlwy5ah9vr2dy4ns8a5fna7wbnq-python3.12-jwt-1.3.1
stripping (with command strip and flags -S -p) in  /nix/store/1ykfhhlwy5ah9vr2dy4ns8a5fna7wbnq-python3.12-jwt-1.3.1/lib
shrinking RPATHs of ELF executables and libraries in /nix/store/mbyzgfa4xwbij1fx4pl6x1y9wmchpy63-python3.12-jwt-1.3.1-dist
checking for references to /build/ in /nix/store/mbyzgfa4xwbij1fx4pl6x1y9wmchpy63-python3.12-jwt-1.3.1-dist...
patching script interpreter paths in /nix/store/mbyzgfa4xwbij1fx4pl6x1y9wmchpy63-python3.12-jwt-1.3.1-dist
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
Running phase: installCheckPhase
@nix { "action": "setPhase", "phase": "installCheckPhase" }
no Makefile or custom installCheckPhase, doing nothing
Running phase: pythonCatchConflictsPhase
@nix { "action": "setPhase", "phase": "pythonCatchConflictsPhase" }
Running phase: pythonRemoveBinBytecodePhase
@nix { "action": "setPhase", "phase": "pythonRemoveBinBytecodePhase" }
Running phase: pythonImportsCheckPhase
@nix { "action": "setPhase", "phase": "pythonImportsCheckPhase" }
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: jwt
Running phase: pytestCheckPhase
@nix { "action": "setPhase", "phase": "pytestCheckPhase" }
Executing pytestCheckPhase
pytest flags: -m pytest
============================= test session starts ==============================
platform linux -- Python 3.12.11, pytest-8.3.5, pluggy-1.6.0
rootdir: /build/source
configfile: setup.cfg
plugins: cov-nixpkgs-stub-1.0.0
collecting ... ^Mcollected 48 items                                                             

jwt/tests/test_jwa.py ....                                               [  8%]
jwt/tests/test_jwk.py .......F................                           [ 58%]
jwt/tests/test_jwkset.py ...                                             [ 64%]
jwt/tests/test_jws.py ...                                                [ 70%]
jwt/tests/test_jwt.py .......                                            [ 85%]
jwt/tests/test_utils.py .......                                          [100%]

=================================== FAILURES ===================================
______________________________ test_jwk_from_der _______________________________

content = b'0\x82\x04\xa4\x02\x01\x00\x02\x82\x01\x01\x00\xb4\xcf\xd1^3)\xec\x0b\xcf\xaev\xf5\xfe-\xc8\x99\xc6xy\xb9\x18\xf8\x0b...7X\xc4\x9d\x1a\xa9\x03\x01\x1d\xc3\xb4\xbe\xf2]2W&\x8d\x92n\xbay\x12<\xd2\x9f\xb0>
private_loader = <built-in function load_der_private_key>

    @jwk_from_bytes_argument_conversion
    def jwk_from_private_bytes(
        content: bytes,
        private_loader: PrivateKeyLoaderT,
        *,
        password: Optional[str] = None,
        backend: Optional[object] = None,
        options: Optional[Mapping[str, object]] = None,
    ) -> AbstractJWKBase:
        """This function is meant to be called from jwk_from_bytes"""
        if options is None:
            options = {}
        try:
>           privkey = private_loader(content, password, backend)  # type: ignore[operator]  # noqa: E501
E           ValueError: Could not deserialize key data. The data may be in an incorrect format, it may be encrypted with an unsupported algorithm, or it may be an unsupported key type (e.g. EC curves with explici>

jwt/jwk.py:345: ValueError

The above exception was the direct cause of the following exception:

content = b'0\x82\x04\xa4\x02\x01\x00\x02\x82\x01\x01\x00\xb4\xcf\xd1^3)\xec\x0b\xcf\xaev\xf5\xfe-\xc8\x99\xc6xy\xb9\x18\xf8\x0b...7X\xc4\x9d\x1a\xa9\x03\x01\x1d\xc3\xb4\xbe\xf2]2W&\x8d\x92n\xbay\x12<\xd2\x9f\xb0>
private_loader = 'load_der_private_key', public_loader = 'load_der_public_key'

    def jwk_from_bytes(
        content: bytes,
        private_loader: PrivateKeyLoaderT,
        public_loader: PublicKeyLoaderT,
        *,
        private_password: Optional[str] = None,
        backend: Optional[object] = None,
        options: Optional[Mapping[str, object]] = None,
    ) -> AbstractJWKBase:
        try:
>           return jwk_from_private_bytes(
                content,
                private_loader,
                password=private_password,
                backend=backend,
                options=options,
            )

jwt/jwk.py:384: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jwt/jwk.py:328: in wrapper
    return func(content, loader, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

content = b'0\x82\x04\xa4\x02\x01\x00\x02\x82\x01\x01\x00\xb4\xcf\xd1^3)\xec\x0b\xcf\xaev\xf5\xfe-\xc8\x99\xc6xy\xb9\x18\xf8\x0b...7X\xc4\x9d\x1a\xa9\x03\x01\x1d\xc3\xb4\xbe\xf2]2W&\x8d\x92n\xbay\x12<\xd2\x9f\xb0>
private_loader = <built-in function load_der_private_key>

    @jwk_from_bytes_argument_conversion
    def jwk_from_private_bytes(
        content: bytes,
        private_loader: PrivateKeyLoaderT,
        *,
        password: Optional[str] = None,
        backend: Optional[object] = None,
        options: Optional[Mapping[str, object]] = None,
    ) -> AbstractJWKBase:
        """This function is meant to be called from jwk_from_bytes"""
        if options is None:
            options = {}
        try:
            privkey = private_loader(content, password, backend)  # type: ignore[operator]  # noqa: E501
            if isinstance(privkey, RSAPrivateKey):
                return RSAJWK(privkey, **options)
            raise UnsupportedKeyTypeError('unsupported key type')
        except ValueError as ex:
>           raise UnsupportedKeyTypeError('this is probably a public key') from ex
E           jwt.exceptions.UnsupportedKeyTypeError: this is probably a public key

jwt/jwk.py:350: UnsupportedKeyTypeError

During handling of the above exception, another exception occurred:

content = b'0\x82\x04\xa4\x02\x01\x00\x02\x82\x01\x01\x00\xb4\xcf\xd1^3)\xec\x0b\xcf\xaev\xf5\xfe-\xc8\x99\xc6xy\xb9\x18\xf8\x0b...7X\xc4\x9d\x1a\xa9\x03\x01\x1d\xc3\xb4\xbe\xf2]2W&\x8d\x92n\xbay\x12<\xd2\x9f\xb0>
public_loader = <built-in function load_der_public_key>

    @jwk_from_bytes_argument_conversion
    def jwk_from_public_bytes(
        content: bytes,
        public_loader: PublicKeyLoaderT,
        *,
        backend: Optional[object] = None,
        options: Optional[Mapping[str, object]] = None
    ) -> AbstractJWKBase:
        """This function is meant to be called from jwk_from_bytes"""
        if options is None:
            options = {}
        try:
>           pubkey = public_loader(content, backend)  # type: ignore[operator]
E           ValueError: Could not deserialize key data. The data may be in an incorrect format, it may be encrypted with an unsupported algorithm, or it may be an unsupported key type (e.g. EC curves with explici>

jwt/jwk.py:365: ValueError

The above exception was the direct cause of the following exception:

    def test_jwk_from_der():
>       jwk_priv = jwk_from_der(load_testdata('rsa_privkey.der'))

jwt/tests/test_jwk.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jwt/jwk.py:420: in jwk_from_der
    return jwk_from_bytes(
jwt/jwk.py:392: in jwk_from_bytes
    return jwk_from_public_bytes(
jwt/jwk.py:328: in wrapper
    return func(content, loader, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

content = b'0\x82\x04\xa4\x02\x01\x00\x02\x82\x01\x01\x00\xb4\xcf\xd1^3)\xec\x0b\xcf\xaev\xf5\xfe-\xc8\x99\xc6xy\xb9\x18\xf8\x0b...7X\xc4\x9d\x1a\xa9\x03\x01\x1d\xc3\xb4\xbe\xf2]2W&\x8d\x92n\xbay\x12<\xd2\x9f\xb0>
public_loader = <built-in function load_der_public_key>

    @jwk_from_bytes_argument_conversion
    def jwk_from_public_bytes(
        content: bytes,
        public_loader: PublicKeyLoaderT,
        *,
        backend: Optional[object] = None,
        options: Optional[Mapping[str, object]] = None
    ) -> AbstractJWKBase:
        """This function is meant to be called from jwk_from_bytes"""
        if options is None:
            options = {}
        try:
            pubkey = public_loader(content, backend)  # type: ignore[operator]
            if isinstance(pubkey, RSAPublicKey):
                return RSAJWK(pubkey, **options)
            raise UnsupportedKeyTypeError(
                'unsupported key type')  # pragma: no cover
        except ValueError as why:
>           raise UnsupportedKeyTypeError('could not deserialize') from why
E           jwt.exceptions.UnsupportedKeyTypeError: could not deserialize

jwt/jwk.py:371: UnsupportedKeyTypeError
=========================== short test summary info ============================
FAILED jwt/tests/test_jwk.py::test_jwk_from_der - jwt.exceptions.UnsupportedKeyTypeError: could not deserialize
========================= 1 failed, 47 passed in 2.08s =========================
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.15.2-cachyos, NixOS, 25.11 (Xantusia), 25.11.20250617.9e83b64`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.3`
 - nixpkgs: `/nix/store/ny8c07vsrfwcb1c4i3jcpbi3qi4w9wy6-source`

### Notify maintainers

@thornycrackers 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/jwt/default.nix
+++ b/pkgs/development/python-modules/jwt/default.nix
@@ -3,6 +3,7 @@
   buildPythonPackage,
   fetchFromGitHub,
   pythonOlder,
+  setuptools,
   cryptography,
   freezegun,
   pytestCheckHook,
@@ -12,11 +13,11 @@
 buildPythonPackage rec {
   pname = "jwt";
   version = "1.4.0";
-  format = "setuptools";
+  pyproject = true;
+
   disabled = pythonOlder "3.6";
 
   src = fetchFromGitHub {
-    inherit version;
     owner = "GehirnInc";
     repo = "python-jwt";
     tag = "v${version}";
@@ -28,7 +29,9 @@ buildPythonPackage rec {
     substituteInPlace setup.cfg --replace "--flake8" ""
   '';
 
-  propagatedBuildInputs = [ cryptography ];
+  build-system = [ setuptools ];
+
+  dependencies = [ cryptography ];
 
   nativeCheckInputs = [
     pytestCheckHook
@@ -38,10 +41,10 @@ buildPythonPackage rec {
 
   pythonImportsCheck = [ "jwt" ];
 
-  meta = with lib; {
+  meta = {
     description = "JSON Web Token library for Python 3";
     homepage = "https://github.com/GehirnInc/python-jwt";
-    license = licenses.asl20;
-    maintainers = with maintainers; [ thornycrackers ];
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ thornycrackers ];
   };
 }

--- a/pkgs/development/python-modules/jwt/default.nix
+++ b/pkgs/development/python-modules/jwt/default.nix
@@ -11,7 +11,7 @@
 
 buildPythonPackage rec {
   pname = "jwt";
-  version = "1.3.1";
+  version = "1.4.0";
   format = "setuptools";
   disabled = pythonOlder "3.6";
 
@@ -19,8 +19,8 @@ buildPythonPackage rec {
     inherit version;
     owner = "GehirnInc";
     repo = "python-jwt";
-    rev = "v${version}";
-    hash = "sha256-N1J8yBVX/O+92cRp+q2gA2cFsd+C7JjUR9jo0VGoINg=";
+    tag = "v${version}";
+    hash = "sha256-Cv64SmhkETm8mx1Kj5u0WZpCPjPNvC+KS6/XaMzxCho=";
   };
 
   postPatch = ''

